zipkin/src/test/java/com/linecorp/armeria/it/tracing/HttpTracingNestedContextIntegrationTest.java

```diff
@@ -0,0 +1,184 @@
+import com.linecorp.armeria.server.tracing.HttpTracingService;
+import com.linecorp.armeria.testing.server.ServerRule;
+
+import brave.Tracing;
+import brave.propagation.CurrentTraceContext;
+import brave.sampler.Sampler;
+import zipkin2.Span;
+
+public class HttpTracingNestedContextIntegrationTest {
+
+    private static final ReporterImpl spanReporter = new ReporterImpl();
+
+    private HttpClient poolHttpClient;
+
+    private final CountDownLatch waitCreateCache = new CountDownLatch(1);
+
+    @Rule
+    public final ServerRule server = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+
+            final CountDownLatch countDownLatch = new CountDownLatch(2);
+            final AtomicReference<CompletableFuture<HttpStatus>> cache = new AtomicReference<>();
+
+            sb.service("/non-trace", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                        throws Exception {
+                    if (Tracing.currentTracer().currentSpan() != null) {
+                        return HttpResponse.of(INTERNAL_SERVER_ERROR);
+                    }
+                    return HttpResponse.of(OK);
+                }
+            });
+
+            sb.service("/create-cache", decorate("service/create-cache", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                        throws Exception {
+                    final CompletableFuture<HttpStatus> future = CompletableFuture.supplyAsync(() -> {
+                        try {
+                            countDownLatch.await(10, TimeUnit.SECONDS);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        return OK;
+                    }, RequestContext.current().contextAwareEventLoop());
+                    cache.set(future);
+                    waitCreateCache.countDown();
+                    return HttpResponse.from(future.thenApply(status -> {
+                        if (Tracing.currentTracer().currentSpan() == null) {
+                            return HttpResponse.of(INTERNAL_SERVER_ERROR);
+                        }
+                        return HttpResponse.of(status);
+                    }));
+                }
+            }));
+
+            sb.service("/read-cache", decorate("service/read-cache", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+                    try {
+                        final RequestContext requestContext = RequestContext.current();
+                        return HttpResponse.from(
+                                cache.get().thenApply(status -> {
```

Comment:
Or the application dev must always know using `thenApplyAsync` with current evetloop.